<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Проверка-данных-на-аномалии-и-исправления.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Заполнение-пропусков" data-toc-modified-id="Заполнение-пропусков-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Заполнение пропусков</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Удаление-дубликатов" data-toc-modified-id="Удаление-дубликатов-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Удаление дубликатов</a></span></li><li><span><a href="#Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Категоризация-дохода" data-toc-modified-id="Категоризация-дохода-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Категоризация дохода</a></span></li><li><span><a href="#Категоризация-целей-кредита" data-toc-modified-id="Категоризация-целей-кредита-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Категоризация целей кредита</a></span></li><li><span><a href="#Категоризация-количества-детей" data-toc-modified-id="Категоризация-количества-детей-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Категоризация количества детей</a></span><ul class="toc-item"><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-9.1.0.1"><span class="toc-item-num">9.1.0.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-1:" data-toc-modified-id="Вывод-1:-9.1.0.2"><span class="toc-item-num">9.1.0.2&nbsp;&nbsp;</span>Вывод 1:</a></span></li><li><span><a href="#Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2:-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-9.1.0.3"><span class="toc-item-num">9.1.0.3&nbsp;&nbsp;</span>Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-2:" data-toc-modified-id="Вывод-2:-9.1.0.4"><span class="toc-item-num">9.1.0.4&nbsp;&nbsp;</span>Вывод 2:</a></span></li><li><span><a href="#Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3:-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-9.1.0.5"><span class="toc-item-num">9.1.0.5&nbsp;&nbsp;</span>Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вывод-3:" data-toc-modified-id="Вывод-3:-9.1.0.6"><span class="toc-item-num">9.1.0.6&nbsp;&nbsp;</span>Вывод 3:</a></span></li><li><span><a href="#Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4:-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-9.1.0.7"><span class="toc-item-num">9.1.0.7&nbsp;&nbsp;</span>Вопрос 4: Как разные цели кредита влияют на его возврат в срок?</a></span></li><li><span><a href="#Вывод-4:" data-toc-modified-id="Вывод-4:-9.1.0.8"><span class="toc-item-num">9.1.0.8&nbsp;&nbsp;</span>Вывод 4:</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

# Исследование надежности заемщиков

Цель исследования - проверить четыре гипотезы:

* Есть ли зависимость между количеством детей и возвратом кредита в срок?
* Есть ли зависимость между семейным положением и возвратом кредита в срок?
* Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
* Как разные цели кредита влияют на его возврат в срок?

Входные данные от банка — статистика о платёжеспособности клиентов.

## Обзор данных

Импорт библиотеки `pandas`:

In [1]:
import pandas as pd

Чтение файла `data.csv` из папки `/datasets` и сохранение его в переменной `df`:

In [2]:
df = pd.read_csv('/datasets/data.csv')

Вывод первых 15 строк таблицы `df`:

In [3]:
df.head(15)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   

     total_income                            purpose  
0   253875.639453                      покупка жилья  
1   112080.014102            приобретение автомобиля  
2   145885.952297                      покупка жилья  
3   267628.550329         дополнительное образование  
4   158616.077870                    сыграть свадьбу  
5   255763.565419                      покупка жилья  
6   240525.971920                  операции с жильем  
7   135823.934197                        образование  
8    95856.832424              на проведение свадьбы  
9   144425.938277            покупка жилья для семьи  
10  113943.491460               покупка недвижимости  
11   77069.234271  покупка коммерческой недвижимости  
12            NaN                    сыграть свадьбу  
13  130458.228857            приобретение автомобиля  
14  165127.911772         покупка жилой недвижимости

Получение общей информации о таблице `df`:

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице двенадцать столбцов, согласно документации к данным:

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита




Выводы по обзору данных.

По обзору первых 10 строчек видно, что в данных есть аномальные значения в столбце `days_employed` -  в следующем шаге 2.1  будет проводиться проверка на аномалии и их исправление, разный регистр в столбце `education`, что приводит к дубликатам и будет исправляться в шаге 2.4.

Благодаря методу info() было выявлено, что количество значений в столбцах различается, в `days_employed` и `total_income` значения меньше общего числа строк. Значит, в этих столбцах есть пропущенные значения, которые нужно обработать.

Во входных данных есть все значения, которые нужны для проверки заданных гипотез.

## Проверка данных на аномалии и исправления.

Проверим необходимые для проверки гипотез столбцы на наличие аномальных значений: `children`,  `family_status`, `income_type`, `purpose`, `debt`.

И столбцы: `gender`, `dob_years`, `days_employed` (потому что были видны минусовые значения, которых не должно быть в трудовом стаже).

In [5]:
df['children'].value_counts() #  20 детей и -1

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

По столбцу с количеством детей в семье выделяются значения 20 и -1. Скорее всего 20 это опечатка от 2, а значение -1 видимо должно было быть 1, но выгрузилось с минусом. Заменим значения 20 на 2 и -1 на 1:

In [6]:
df.loc[df['children'] == 20, 'children'] = 2

In [7]:
df.loc[df['children'] == -1, 'children'] = 1

In [8]:
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

In [9]:
df['family_status'].value_counts() # нет аномалий

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

In [10]:
df['income_type'].value_counts() # нет аномалий

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

In [11]:
df['purpose'].value_counts()  

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Опечатка "ремонт жилью" 612 раз, будем учитывать её при составлении условий для категорий. 

In [12]:
df['debt'].value_counts() # нет аномалий

0    19784
1     1741
Name: debt, dtype: int64

In [13]:
df['gender'].value_counts() # неизвестное значение XNA

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Столбец с полом клиентов в дальнейшем исследовании нам не нужен. Поэтому не будем его редактировать.

In [14]:
df['dob_years'].value_counts() 

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

В столбце `dob_years` есть возраст 0, но этот столбец не влияет на исследование, следовательно не будем его обрабатывать.

В столбце `days_employed` присутствует такие аномалии как отрицательное количество дней трудового стажа, а для тех данных, в которых это число положительное это не может быть количеством дней трудового стажа, так как это число в 15-20 раз больше продолжительности жизни. Можно предположить, что все положительные значения это количество трудового стажа в часах. Но так как эти данные в дальнейшем исследовании нам не пригодятся, переводить в дни не имеет смысла, ведь возможно это предположение ложное.

Об этих аномалиях необходимо сообщить тем, кто занимается выгрузкой данных или их заполнением в этом столбце. 

Переведем все числа в положительные, используя модуль числа. 

In [15]:
df['days_employed'] = df['days_employed'].abs()

Вывод:
В данных были обнаружены различные аномалии 20 детей и -1 в столбце `children`, неизвестное значение XNA в столбце `gender`, нулевой возраст в столбце `dob_years` - это скорее всего точечные аномалии, которые сильно выбиваются из данных. И самая заметная в столбце `days_employed` это отрицательные дни трудового стажа.

Об этих находках необходимо сообщить сотрудникам, занимающимся выгрузкой или заполнением этих данных.

Причинами могут послужить разные факторы, такие как: человеческий фактор, технические ошибки (ошибки в выгрузке), ошибки в данных (неверная запись). 

## Заполнение пропусков

Подсчёт пропусков в таблице `df`:

In [16]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Таким образом обнаружены пропуски в двух столбцах: общий трудовой стаж в днях и ежемесячный доход клиентов.

Проверим какую долю составляют пропущенные значения в каждом из столбцов с пропусками:

In [17]:
part_days_employed = df['days_employed'].isna().sum() / df['days_employed'].isna().count()
print(' Доля пропущенных значений общего трудового стажа {:.1%}'.format(part_days_employed))

 Доля пропущенных значений общего трудового стажа 10.1%


In [18]:
part_total_income = df['total_income'].isna().sum() / df['total_income'].isna().count()
print(' Доля пропущенных значений ежемесячного дохода {:.1%}'.format(part_total_income))

 Доля пропущенных значений ежемесячного дохода 10.1%


Общий трудовой стаж в днях и ежемесячный доход — количественные переменные. Пропуски в таких переменных заполняют характерными значениями для выборки. Чтобы примерно оценить типичные значения выборки, годятся среднее арифметическое или медиана. Так как в выборке присутствуют выдающиеся значения (340266.072047 против 152.779569) лучше использовать медиану.

Заполним пропуски медианным значением по столбцам:

In [19]:
days_employed_median = df['days_employed'].median()
print(days_employed_median)
df['days_employed'] = df['days_employed'].fillna(value = days_employed_median)
#выведем строки с незаполненными ранее данными (строка 12,26,65), чтобы убедиться, что всё заполнено
df.loc[[12,26,65]]

2194.220566878695


children  days_employed  dob_years education  education_id  \
12         0    2194.220567         65   среднее             1   
26         0    2194.220567         41   среднее             1   
65         0    2194.220567         21   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
65  Не женат / не замужем                 4      M    компаньон     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
65           NaN  операции с коммерческой недвижимостью

Заполняя пропуски просто медианным значением (как я сделала сначала) окажется, что все 10% пропусков будут заполнены доходом в 145017, получается, что они заполнят в дальнейшем одну категорию "C". Поэтому имеет смысл заменить медианным значением по типу занятости.

In [20]:
income_grouped = df.groupby('income_type')['total_income'].median()
income_grouped # табличка для проверки правильно ли я сделала замену, которая в ячейке ниже.

income_type
безработный        131339.751676
в декрете           53829.130729
госслужащий        150447.935283
компаньон          172357.950966
пенсионер          118514.486412
предприниматель    499163.144947
сотрудник          142594.396847
студент             98201.625314
Name: total_income, dtype: float64

In [21]:
# Создадим цикл для каждого элемента в столбце total_income с пропущенными значениями и каждым уникальным
# значением столбца income_type.Зададим две переменные: 
# median - подсчитывающую медиану столбца total_income для заданного типа занятости income_type в столбце income_type
# соединим два условия тип занятости и пропущенное значение столбца total_income и
# заменим наконец-то каждое пропущеное значение дохода медианой по типу занятости.

for income_type in df[df['total_income'].isna()]['income_type'].unique():
    median = df[df['income_type'] == income_type]['total_income'].median()
    df.loc[(df['income_type'] == income_type) & (df['total_income'].isna())] = \
    df.loc[(df['income_type'] == income_type) & (df['total_income'].isna())].fillna(median)
#выведем строки с незаполненными ранее данными (строка 12,26,65,82), чтобы убедиться, что всё хорошо
df.loc[[12,26,65,82]]


children  days_employed  dob_years education  education_id  \
12         0    2194.220567         65   среднее             1   
26         0    2194.220567         41   среднее             1   
65         0    2194.220567         21   среднее             1   
82         2    2194.220567         50    высшее             0   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
65  Не женат / не замужем                 4      M    компаньон     0   
82        женат / замужем                 0      F    сотрудник     0   

     total_income                                purpose  
12  118514.486412                        сыграть свадьбу  
26  150447.935283                            образование  
65  172357.950966  операции с коммерческой недвижимостью  
82  142594.396847                                  жилье

Проверка на наличие пропусков в таблице df

In [23]:
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Пропуски в столбцах "общий трудовой стаж в днях" и "ежемесячный доход" на первый взгляд могут быть обусловлены отсутствием официального трудоустройства, но каким образом банк выдал кредит лицу, который даже не предоставил документы, подтверждающие финансовое состояние и трудовую занятость заемщика? Возможно у клиентов есть поручители или они готовы оставить в залог имущество. В общем необходимо уточнить эту информацию у сотрудников банка, возможно банк не проверяет у клиентов официальный доход, но тогда почему он есть у других клиентов? Также возможно, если у банка много филиалов, из какого-то из них произошел сбой в выгрузке данных или просто эти данные не заполнил сотрудник. 


## Изменение типов данных

Замена вещественного типа данных в столбце `total_income` на целочисленный методом `astype()`.

In [24]:
try:
    df['total_income'] = df['total_income'].astype('int')
except:
    print('Ошибочка')
df.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

Заменим и в столбце `days_employed`:

In [25]:
try:
    df['days_employed'] = df['days_employed'].astype('int')
except:
    print('Ошибочка')
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

## Удаление дубликатов

Поиск явных дубликатов и подсчет их числа:

In [26]:
df.duplicated().sum()

54

Удалим дубликаты, создадим новые индексы, удаляя старые:

In [27]:
df = df.drop_duplicates().reset_index(drop=True)

Проверка 

In [28]:
df.duplicated().sum()

0

Проверка неявных дубликатов:

In [29]:
df['education'].value_counts() # разный регистр 

среднее                13705
высшее                  4710
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   273
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

По результату работы функции value_counts() можно сделать вывод о  том, что дубликаты появились в следствии разного регистра букв. Исправим это используя метод str.lower():

In [30]:
df['education'] = df['education'].str.lower()
df['education'].unique()

array(['высшее', 'среднее', 'неоконченное высшее', 'начальное',
       'ученая степень'], dtype=object)

Для поиска явных дубликатов я использовала метод `duplicated()` в сочетании с методом `sum()`, потому что он возвращает количество дублирующих строк. И с помощью метода `drop_duplicates().reset_index(drop=True)` удалила эти повторяющиеся строки.

Для проверки неявных дубликатов, я использовала метод `value_counts()`, который анализирует столбец, выбирает каждое уникальное значение и подсчитывает частоту его встречаемости в столбце `education`, таким образом было выявлено, что одинаковые значения были записаны разным регистром. Для приведения к одному регистру я использовала метод `str.lower()`.

Наши неявные дубликаты это альтернативные написания одного и того же значения, т.е. при заполнении данных возможно у кого-то был включен Caps Lock, кто-то решил заполнять данные с большой буквы. Скорее всего причина таких дубликатов человеческий фактор, отсутствие правила по заполнению этой колонки. Причиной же явных повторений одних и тех же строк может быть задвоение данных.

## Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим новые датафреймы словари для столбцов `education` и `family_status` для того, чтобы обращаться к ним в дальнейшем по идентификатору. 

In [31]:
education_dict = df[['education_id','education']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True) # Удалим дубликаты
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [32]:
status_dict = df[['family_status_id','family_status']]
status_dict = status_dict.drop_duplicates().reset_index(drop=True) # Удалим дубликаты
status_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  Не женат / не замужем

Удаляем два столбца и заменяем исходный df:

In [33]:
df.drop(columns = ['education', 'family_status'],axis = 1, inplace=True)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  
0   сотрудник     0        253875               покупка жилья  
1   сотрудник     0        112080     приобретение автомобиля  
2   сотрудник     0        145885               покупка жилья  
3   сотрудник     0        267628  дополнительное образование  
4   пенсионер     0        158616             сыграть свадьбу

Были созданы два словаря education_dict и status_dict, удалены лишние столбцы `education` и `family_status`. Теперь к словарям можно обращаться по идентификатору education_id и family_status_id.

## Категоризация дохода

 Для категоризации дохода напишем функцию.
 На вход функции помещается среднемесячный доход, а возвращает она категорию, используя правила:
* 0–30000 — 'E';
* 30001–50000 — 'D';
* 50001–200000 — 'C';
* 200001–1000000 — 'B';
* 1000001 и выше — 'A'.

Далее протестируем функцию значениями из разных категорий. И добавим в первоначальную таблицу с данными отдельный столбец total_income_category с этими категориями.

In [34]:
def income_category(total_income):
    if 0 < total_income < 30000:
        return 'E'
    if 30001 < total_income < 50000:
        return 'D'
    if 50001 < total_income < 200000:
        return 'C'
    if 200001 < total_income < 1000000:
        return 'B'
    return 'A'
print(income_category(25000))
print(income_category(235000))



E
B


In [35]:
df['total_income_category'] = df['total_income'].apply(income_category)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C

Чтобы ознакомиться с данными по категориям выведем по ним статистику методом value_counts():

In [36]:
df['total_income_category'].value_counts()

C    16032
B     5041
D      350
A       26
E       22
Name: total_income_category, dtype: int64

Вывод: 
* Основными клиентами банка являются люди, попадающие в категории "B" и "C" с доходом от 50тыс до 1млн.Это может объясняться тем, что у них есть высокий стабильный доход, они уверены в том, что смогут выплатить кредит, их не останаливают проценты по нему и они хотят получить то ради чего они берут кредит сейчас, а не копить на свои цели. 
* Клиентов категории "E" и"A" с доходом меньше 30тыс и больше 1 млн одинаково мало. Также стоит отметить клиетов категории "D" с доходом до 50 тыс, которых тоже мало относительно категорий "B" и "C" в данной выборке. Причины у клиентов с доходом меньше 50 тыс скорее всего обратные вышеперечисленным, т.е. невысокий доход, относительно которого проценты по кредиту кажутся слишком большими и в целях экономии люди копят на свои цели. А клиентам с доходом более 1 млн видимо брать кредит не имеет смысла. 

## Категоризация целей кредита

Для начала изучим данные в столбце purpose:

In [37]:
df['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создадим функцию которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:
* операции с автомобилем,
* операции с недвижимостью,
* проведение свадьбы,
* получение образования.

In [38]:
def for_purpose_category(purpose):
    if 'автомобил' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose or 'недвиж' in purpose:
        return 'операции с недвижимостью'
    if 'свадьб' in purpose:
        return 'проведение свадьбы'
    if 'образован' in purpose:
        return 'получение образования'
    return 'другое' # чтобы проверить не упустила ли я что-то.
print(for_purpose_category('покупка жилой недвижимости')) # проверки работы функции
print(for_purpose_category('сделка с автомобилем'))


операции с недвижимостью
операции с автомобилем


In [39]:
df['purpose_category'] = df['purpose'].apply(for_purpose_category)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

In [40]:
df['purpose_category'].value_counts()

операции с недвижимостью    10814
операции с автомобилем       4308
получение образования        4014
проведение свадьбы           2335
Name: purpose_category, dtype: int64

Операция по категоризации целей кредита прошла успешно. 

* Больше всего клиентов берут кредит с целью проведения операций с недвижимостью (10814).
* На операции с автомобилем (4308) и получение образования (4014) приходится практически одинаковое число клиентов.
* Меньше всего клиентов берут кредит с целью проведения свадьбы (2335).

## Категоризация количества детей

Создадим функцию для категоризации клиентов с детьми, чтобы использовать её в дальнейшем при проверки гипотезы о зависимоти количества детей и возвратом кредита в срок.

In [41]:
def children_category (children):
    if children == 0:
        return 'нет детей'
    if children == 1:
        return 'один ребёнок'
    if children == 2:
        return 'два ребёнка'
    return 'от 3 до 5 детей'
print(children_category(2)) # проверки работы функции
print(children_category(5))


два ребёнка
от 3 до 5 детей


In [42]:
df['children_category'] = df['children'].apply(children_category)
df.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      F   
1         1           4024         36             1                 0      F   
2         0           5623         33             1                 0      M   
3         3           4124         32             1                 0      M   
4         0         340266         53             1                 1      F   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   

  total_income_category          purpose_category children_category  
0                     B  операции с недвижимостью      один ребёнок  
1                     C    операции с автомобилем      один ребёнок  
2                     C  операции с недвижимостью         нет детей  
3                     B     получение образования   от 3 до 5 детей  
4                     C        проведение свадьбы         нет детей

Итого: мы добились того, что семьи с количеством детей больше трех выводятся в отдельную категорию для того, чтобы сделать выводы на основании большего количества клиентов т.к. если рассматривать эти группы по отдельности сложно выявить какую-то зависимость.

### Ответы на вопросы.

Чтобы отталкиваться от какого-то значения и принимать решение в дальнейшем есть ли зависимость от каких-либо данных, сперва подсчитаем общий процент невыплаченного вовремя кредита:

In [43]:
late_payment = df[(df['debt'] == 1)].count() / df['debt'].count()
'Доля невыплаченного в срок кредита равна {:.2%}'.format(late_payment['debt'])

'Доля невыплаченного в срок кредита равна 8.11%'

##### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

Создаём сводную таблицу. Функцией sum() подсчитаем все значения с 1 в столбце debt, функцией count() подсчитаем сколько всего значений в столбце. Добавим столбец ratio, отражающий долю клиентов, невыплативших кредит в срок в общем числе клиентов.

In [44]:
data_pivot = df.pivot_table(index='children_category', values ='debt', aggfunc=['sum', 'count'] )
data_pivot['ratio'] = data_pivot['sum']/data_pivot['count'] * 100
data_pivot.sort_values(by= 'ratio')

sum  count     ratio
                   debt   debt          
children_category                       
нет детей          1063  14107  7.535266
от 3 до 5 детей      31    380  8.157895
один ребёнок        445   4856  9.163921
два ребёнка         202   2128  9.492481

##### Вывод 1:

* На основании данных можно сдеть вывод о том, что клиенты у которых нет детей (7,53%)чаще остальных выплачивают кредиты в срок.
* Клиенты у которых больше трех детей (8.15%) выплачивают кредит практически так же как и основная масса.
* Клиенты с одним (9.16%) и двумя детьми (9.49%) чаще выплачивают кредиты позже срока.

##### Вопрос 2: Есть ли зависимость между семейным положением и возвратом кредита в срок?

Делаем то же самое, что и при ответе на первый вопрос, только добавим столбец `family_status` методом .insert(), чтобы не вспоминать какой id какому статусу соответствует. И сделаем сортировку. 

In [45]:
data_pivot2 = df.pivot_table(index='family_status_id', values ='debt', aggfunc=['sum', 'count'] )
data_pivot2['ratio'] = data_pivot2['sum']/data_pivot2['count'] * 100
data_pivot2.insert(0, 'family_status', status_dict.get('family_status'), allow_duplicates = False)
data_pivot2.sort_values(by= 'ratio')

family_status  sum  count     ratio
                                        debt   debt          
family_status_id                                             
2                        вдовец / вдова   63    959  6.569343
3                             в разводе   85   1195  7.112971
0                       женат / замужем  931  12344  7.542126
1                      гражданский брак  388   4163  9.320202
4                 Не женат / не замужем  274   2810  9.750890

##### Вывод 2:

Выделим категории, в которых клиентов с просроченным кредитом относительно меньше:
* вдовец / вдова (6.56%)
* в разводе	(7.11%)
* женат / замужем (7.54%)

Остальные клиенты, состоящие в гражданском браке (9.3%) и не женатые (9.75%) чаще нарушают сроки выплаты.

##### Вопрос 3: Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [46]:
data_pivot3 = df.pivot_table(index='total_income_category', values = 'debt', aggfunc= ['sum', 'count'])
data_pivot3['ratio'] = data_pivot3['sum']/data_pivot3['count'] * 100
data_pivot3.sort_values(by= 'ratio')

sum  count     ratio
                       debt   debt          
total_income_category                       
D                        21    350  6.000000
B                       356   5041  7.062091
A                         2     26  7.692308
C                      1360  16032  8.483034
E                         2     22  9.090909

##### Вывод 3:

* Клиентов в категориях "B" и "C" большинство. Меньше всего клиентов категори "A" и "E".
* Судя по данным в срок укладываются чаще клиенты с категориями дохода "A" и "B". 
* Доля невыплаченных кредитов в срок у категорий "C" и "E" выше чем у остальных.

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [47]:
data_pivot4 = df.pivot_table(index='purpose_category', values ='debt', aggfunc=['sum', 'count'] )
data_pivot4['ratio'] = data_pivot4['sum']/data_pivot4['count'] * 100
data_pivot4.sort_values(by= 'ratio')

sum  count     ratio
                         debt   debt          
purpose_category                              
операции с недвижимостью  782  10814  7.231367
проведение свадьбы        186   2335  7.965739
получение образования     370   4014  9.217738
операции с автомобилем    403   4308  9.354689

##### Вывод 4:

Клиенты, которые берут кредит на операции с недвижимостью (7.23%) и на проведение свадьбы (7.96%) реже нарушают сроки выплаты, чем те клиенты чья цель кредита - получение образования (9.21%) и на оперции с автомобилем (9.35%).

## Общий вывод:

Подведем итог и составим список-портрет клиентов, которые вероятнее всего выплатят и не выплатят кредит в срок.

Выплатят вовремя:
* клиенты у которых нет детей (7,53%);
* клиенты с семейным положением : вдовец / вдова (6.56%) , в разводе (7.11%), женат / замужем (7.54%);
* клиенты с категориями дохода "A" и "B";
* клиенты, которые берут кредит на операции с недвижимостью (7.23%) и на проведение свадьбы (7.96%)

Чаще не выплачивают в срок:
* клиенты с одним (9.16%) и двумя детьми (9.49%);
* клиенты, состоящие в гражданском браке (9.3%) и не женатые (9.75%);
* клиенты с категориями дохода "C" и "E";
* клиенты чья цель кредита - получение образования (9.21%) и на оперции с автомобилем (9.35%).